c2rust-analyze/src/rewrite/expr/mir_op.rs

```diff
@@ -801,6 +958,58 @@ where
             return Ok(());
         }
 
+        if from.option && from.own != to.own {
+            // Downgrade ownership before unwrapping the `Option` when possible.  This can avoid
+            // moving/consuming the input.  For example, if the `from` type is `Option<Box<T>>` and
+            // `to` is `&mut T`, we start by calling `p.as_mut().as_deref()`, which gives
+            // `Option<&mut T>` without consuming `p`.
+            if !from.own.is_copy() {
+                match to.own {
+                    Ownership::Raw | Ownership::Imm => {
+                        (self.emit)(RewriteKind::OptionDowngrade {
+                            mutbl: false,
+                            deref: true,
+                        });
+                        from.own = Ownership::Imm;
+                    }
+                    Ownership::RawMut | Ownership::Cell | Ownership::Mut => {
+                        (self.emit)(RewriteKind::OptionDowngrade {
+                            mutbl: true,
+                            deref: true,
+                        });
+                        from.own = Ownership::Mut;
+                    }
+                    _ => {
+                        // Remaining cases are unsupported.
```

Comment:
We could log this.

Reply:
I updated this comment to clarify: 73652be